File: 2022-02/2022-2-8(화)/junyoung/1660/1660.cpp
#include <bits/stdc++.h>
using namespace std;

int N;

int main(void) {
    cin >> N;
    vector<int> arr = {1};
    int cnt = 1;
    while(arr.back() <= N) {
        cnt++;
        arr.push_back(arr.back() + cnt);
    }
    
    vector<int> num = {arr[0]};
    for (int i = 1; i < arr.size(); i++) {
        arr[i] += arr[i - 1];
        if (arr[i] > N) break;
        num.push_back(arr[i]);
    }
    
    vector<int> dp(N + 1, 300000);
    dp[0] = 0;
    for (auto &n: num) {
        for (int i = n; i <= N; i++) {
            dp[i] = min((dp[i - n] + 1), dp[i]);
        }
    }
    cout << dp[N];
}

File: 2022-02/2022-2-8(화)/junyoung/1660/1660.py
def f(N):
	dp = [300000] * (N + 1)
	dp[0] = 0

	for n in arr:
		for j in range(n, N + 1):
			dp[j] = min(dp[j - n] + 1, dp[j])

	print(dp[-1])



N = int(input())

arr = [1]
cnt = 1
while arr[-1] <= int(N / 2):
	cnt += 1
	arr.append(arr[-1] + cnt)

for i in range(1, len(arr)):
	arr[i] += arr[i - 1]
	if arr[i] > N:
		arr = arr[:i]
		break

f(N)
